newrelic/resource_newrelic_alert_condition.go

```diff
@@ -191,7 +191,7 @@ func resourceNewRelicAlertCondition() *schema.Resource {
 			"user_defined_value_function": {
 				Type:         schema.TypeString,
 				Optional:     true,
-				ValidateFunc: validation.StringInSlice([]string{"average", "min", "max", "total", "sample_size"}, false),
+				ValidateFunc: validation.StringInSlice([]string{"average", "min", "max", "total", "sample_size", ""}, false),
```

Comment:
Hi there, thanks for the contribution! Did you happen to run into a scenario that needed the empty string? Since `user_defined_value_function` is an optional attribute, it can be omitted, but maybe there's something else going on that we need to address 🙂

Reply:
Hi sander,

Im trying to keep the module as generic as possible so that we can reuse the same module for the terraforming alers which doesnt have user_defined_value

Reply:
Terraform doesnt allow to set up null value as default. so this is a blocker from using a generic module for the creating apm alerts.

Reply:
Its the same with entities. 

Entities should be allowed to have an empty arrray. 
The below output is from restapi of an alert with custom metric. 
```
    {
      "id": 10539306,
      "type": "apm_app_metric",
      "name": "abc Error too high",
      "enabled": true,
      "entities": [],
      "metric": "user_defined",
      "condition_scope": "application",
      "terms": [
        {
          "duration": "5",
          "operator": "above",
          "priority": "critical",
          "threshold": "1",
          "time_function": "any"
        }
      ],
      "user_defined": {
        "metric": "Custom/abc.Error",
        "value_function": "total"
      }
    }
 
```

Reply:
Gotcha, I think I see what you're saying. Could you provide your Terraform module and associated config code? We'll need to attempt to reproduce what you're experiencing 🙂

Reply:
module

```
resource "newrelic_alert_condition" "apm" {
  policy_id = "${var.policy_id}"

  name            = "${var.name}"
  enabled         = "${var.enabled}"
  type            = "${var.type}"
  entities        = "${var.entities}"
  metric          = "${var.metric}"
  runbook_url     = "${var.runbook_url}"
  condition_scope = "${var.condition_scope}"

  term {
    duration      = "${var.duration}"
    operator      = "${var.operator}"
    priority      = "${var.priority}"
    threshold     = "${var.threshold}"
    time_function = "${var.time_function}"
  }

  term {
    duration      = "${var.durationb}"
    operator      = "${var.operatorb}"
    priority      = "${var.priorityb}"
    threshold     = "${var.thresholdb}"
    time_function = "${var.time_function}"
  }
}
```
Variables
```
variable "policy_id" {}

variable "name" {}

variable "enabled" {
  default = "true"
}

variable "entities" {
  type = "list"
}

variable "metric" {}

variable "runbook_url" {
  default = "Default runbook"
}

variable "duration" {}

variable "operator" {
  default = "above"
}

variable "priority" {
  default = "critical"
}

variable "threshold" {}

variable "durationb" {}

variable "operatorb" {
  default = "above"
}

variable "priorityb" {
  default = "warning"
}

variable "condition_scope" {
  default = "application"
}

variable "thresholdb" {}

variable "type" {}

variable "time_function" {
  default = "all"
}

```

Providers
```
module "provider_without_custom_metric" {
  source    = "../../../../modules/newrelic/alertcondition/apm"
  policy_id = "${module.abcd.newrelic_alert_policy_id}"

  name        = "Trigger Failure"
  entities    = [140942965]
  type        = "apm_app_metric"
  metric      = "user_defined"
  runbook_url = ""
  duration    = 60
  operator    = "above"
  priority    = "warning"
  threshold   = "0"
  durationb   = 120
  operatorb   = "above"
  priorityb   = "critical"
  thresholdb  = "0"
}

module "provider_with_custom_metric_without_entity" {
  source    = "../../../../modules/newrelic/alertcondition/apm"
  policy_id = "${module.abcd-policy.newrelic_alert_policy_id}"

  name        = "custom check"
  entities    = []
  type        = "apm_app_metric"
  metric      = "user_defined"
  runbook_url = "some runbook"
  duration    = 5
  operator    = "above"
  priority    = "critical"
  threshold   = "96"
  durationb   = 5
  operatorb   = "above"
  priorityb   = "warning"
  thresholdb  = "96"
  user_defined_metric= "Custom/ticket/status" 
  user_defined_value_function=  "max"
}
```
Ideally, All variables should allow empty string as it gives maximum flexibility to reuse the modules.

Reply:
@soumyajk what version of Terraform are you using? 0.12 added the ability to set [null](https://github.com/hashicorp/terraform/issues/14037) defaults.  Following that pattern in your module would enable any of the optional variables to have `default = null` and work as expected without changes to the provider.

For a good example on this refer to our [APM Module](https://github.com/newrelic/terraform-newrelic-apm).  I've also included a modified version of your example:

### mod/main.tf

```
resource "newrelic_alert_condition" "apm" {
  policy_id = var.policy_id

  name                        = var.name
  enabled                     = var.enabled
  type                        = var.type
  entities                    = var.entities
  metric                      = var.metric
  runbook_url                 = var.runbook_url
  condition_scope             = var.condition_scope
  user_defined_metric         = var.user_defined_metric
  user_defined_value_function = var.user_defined_value_function

  term {
    duration      = var.duration
    operator      = var.operator
    priority      = var.priority
    threshold     = var.threshold
    time_function = var.time_function
  }

  term {
    duration      = var.durationb
    operator      = var.operatorb
    priority      = var.priorityb
    threshold     = var.thresholdb
    time_function = var.time_function
  }
}
```

### mod/variables.tf

```
# Required vars
variable "policy_id" {}
variable "name" {}
variable "type" {}
variable "metric" {}
variable "duration" {}
variable "threshold" {}
variable "durationb" {}
variable "thresholdb" {}

# Defaults
variable "enabled" {
  default = "true"
}

variable "entities" {
  type = "list"
}

variable "runbook_url" {
  default = "Default runbook"
}

variable "operator" {
  default = "above"
}

variable "priority" {
  default = "critical"
}

variable "operatorb" {
  default = "above"
}

variable "priorityb" {
  default = "warning"
}

variable "condition_scope" {
  default = "application"
}

variable "time_function" {
  default = "all"
}

# Optional
variable "user_defined_value_function" {
  default = null
}

variable "user_defined_metric" {
  default = null
}
```

Reply:
Im using 0.11, I will check if i can upgrade to resolve this issue. thanks a lot for coming back on this.

Reply:
Tried this with nrql 

```
resource "newrelic_nrql_alert_condition" "nrql" {
  policy_id = "var.policy_id"

  name                         = "var.name"
  type                         = "var.type"
  runbook_url                  = "var.runbook_url"
  enabled                      = "var.enabled"
  violation_time_limit_seconds = "var.violation_time_limit_seconds"

  term {
    duration      = "var.duration"
    operator      = "var.operator"
    priority      = "var.priority"
    threshold     = "var.threshold"
    time_function = "var.time_function"
  }

  term {
    duration      = "var.durationb"
    operator      = "var.operatorb"
    priority      = "var.priorityb"
    threshold     = "var.thresholdb"
    time_function = "var.time_function"
  }

  nrql {
    query       = "var.query"
    since_value = "var.since_value"
  }

  value_function  = "var.value_function"
  expected_groups = "var.expected_groups"
  ignore_overlap  = "var.ignore_overlap"
}

```

```
variable "policy_id" {}

variable "name" {}

variable "type" {}

variable "enabled" {}

variable "runbook_url" {
  default = "Default runbook"
}

variable "duration" {}

variable "operator" {
  default = "above"
}

variable "priority" {
  default = "critical"
}

variable "threshold" {}

variable "durationb" {
  default = null
}

variable "operatorb" {
  default = "above"
}

variable "priorityb" {
  default = "warning"
}

variable "thresholdb" {
  default = null
}

variable "query" {}

variable "since_value" {}

variable "value_function" {}

variable "time_function" {}

variable "expected_groups" {
  default = null
}

variable "ignore_overlap" {
  default = null
}

variable "violation_time_limit_seconds" {}

```

Error message 

```
Error: Incorrect attribute value type

  on ../../../../modules/newrelic/alertcondition/nrql/monitors.tf line 33, in resource "newrelic_nrql_alert_condition" "nrql":
  33:   ignore_overlap  = "var.ignore_overlap"

Inappropriate value for attribute "ignore_overlap": a bool is required.
```

Reply:
For Terraform 0.12.x, you don't need to quote the values within the module.  Since it is a bool type, I think that the quotes around the var change the default to an empty string (not null)

Try changing the module to the following:

```
resource "newrelic_nrql_alert_condition" "nrql" {
  policy_id = "var.policy_id"

  name                         = var.name
  type                         = var.type
  runbook_url                  = var.runbook_url
  enabled                      = var.enabled
  violation_time_limit_seconds = var.violation_time_limit_seconds

  term {
    duration      = var.duration
    operator      = var.operator
    priority      = var.priority
    threshold     = var.threshold
    time_function = var.time_function
  }

  term {
    duration      = var.durationb
    operator      = var.operatorb
    priority      = var.priorityb
    threshold     = var.thresholdb
    time_function = var.time_function
  }

  nrql {
    query       = var.query
    since_value = var.since_value
  }

  value_function  = var.value_function
  expected_groups = var.expected_groups
  ignore_overlap  = var.ignore_overlap
}
```